transport/cluster.go

```diff
@@ -0,0 +1,268 @@
+	. "scylla-go-driver/frame/response"
+)
+
+var (
+	peerQuery = Statement{
+		Content:     "SELECT peer, data_center, rack, tokens FROM system.peers",
+		Consistency: frame.ONE,
+	}
+
+	localQuery = Statement{
+		Content:     "SELECT rpc_address, data_center, rack, tokens FROM system.local",
+		Consistency: frame.ONE,
+	}
+)
+
+func makePeerStatement(addr frame.Inet) Statement {
+	return Statement{
+		Content:     "SELECT peer, data_center, rack, tokens FROM system.peers WHERE peer = " + net.IP(addr.IP).String(),
+		Consistency: frame.ONE,
+	}
+}
+
+func InetToString(i frame.Inet) string {
+	return string(i.IP) + ":" + strconv.Itoa(int(i.Port))
+}
+
+const (
+	nodeAddr   = 0
+	nodeDC     = 1
+	nodeRack   = 2
+	nodeTokens = 3
+
+	statusDown = 0
+	statusUP   = 1
+
+	closeRefresher  = 0
+	refreshTopology = 1
+)
+
+type PeerMap = map[string]*Node
+
+type Cluster struct {
+	Peers atomic.Value // PeerMap
+
+	Control   ControlConn
+	Refresher chan byte
+
+	// Yet to be used.
+	// Ring        map[Token]*Node
+	// DataCenters map[string]*DataCenter
+}
+
+type DataCenter struct {
+	Nodes   []Node
+	RackCnt uint // Type?
+}
+
+func (c *Cluster) GetPeers() PeerMap {
+	return c.Peers.Load().(PeerMap)
+}
+
+func (c *Cluster) SetPeers(m PeerMap) {
+	c.Peers.Store(m)
+}
+
+// Refresh notifies cluster that it should update topology.
+// Should it wait for refreshing to end?
+func (c *Cluster) Refresh() {
+	c.Refresher <- refreshTopology
+}
+
+func (c *Cluster) StopRefresher() {
+	c.Refresher <- closeRefresher
+}
+
+// RefreshTopology creates new PeerMap filled with the result of both localQuery and peerQuery.
+// The old map is replaced with the new one atomically to prevent dirty reads.
+func (c *Cluster) RefreshTopology() error {
+	peerRes, err := c.Control.Conn.Query(peerQuery, nil)
+	if err != nil {
+		return fmt.Errorf("discovering peer topology: %w", err)
+	}
+
+	localRes, err := c.Control.Conn.Query(localQuery, nil)
+	if err != nil {
+		return fmt.Errorf("discovering local topology: %w", err)
+	}
+
+	rows := append(peerRes.Rows, localRes.Rows[0])
+
+	// If node is present both in both maps we can update already created node
+	// instead of creating new one from the scratch.
+	old := c.GetPeers()
+	m := PeerMap{}
+	for _, v := range rows {
+		addr := net.IP(v[nodeAddr]).String()
+
+		if node, ok := old[addr]; ok {
+			// Even if address is the same
+			// I think that tokens can change over time and should be updated.
+			// Can rack or data center change as well?
+			node.tokens = v[nodeTokens]
+			node.SetStatus(statusUP)
+			m[addr] = node
+		} else {
+			m[addr] = &Node{
+				addr:       addr,
+				datacenter: string(v[nodeDC]),
+				rack:       string(v[nodeRack]),
+				tokens:     v[nodeTokens],
+				pool:       InitNodeConnPool(addr, c.Control.Cfg),
+				status:     statusUP,
+			}
+		}
+	}
+
+	c.SetPeers(m)
+	return nil
+}
+
+type ControlConn struct {
+	Conn       *Conn
+	Cfg        ConnConfig
+	Events     responseHandler
+	KnownHosts []string
+}
```

Comment:
ControlConn can be created from a dialed connection i.e. there should be `NewControlConn(conn Conn)`.

It adds to Conn ability to run some common queries and react to events.

Reply:
Fields must not be exported.